---
title: "Pipeline Todo List"
author: "Brian Davis"
date: "`r Sys.Date()`"
output: html_document
---

Basic Todo list of ideas for the analysis pipeline.

# Basic Pipeline procedure

 * create a phenotype object (basic error checking)
 * create a snpinfo class (basic error checking)
 * create a genotype class (basic error checking)
 * create a gwas_list (reduce pheno/geno/snpinfo into a common minimal data set)
 * create a standard output structure (not part of the package)
 * run arbitrary test
    * seqMeta
      * create a prepScores object
      * run T1/T5/Skat/etc
    * gwaslm
    * firth
 * create reports
    * results tables
    * qq plots
    * manhattan plots
    * compare tests
 * ideas
    * for a gwas_list run an arbitrary function on each "gene" (do??)
    * tidyr / dplyr concepts
 * functions
    * gwaslm
    * maf
    * impute(x, type="mean")
    * cov
    * score
    * SEE seqMeta function

# phenotype

 * create a phenotype class
    * drop unwanted / unneeded columns
 * improved error checking
 * summary function
 * print (similar to data_frame)
 * data_frame support

# genotype

 * create a genotype class
 * improved error checking
 * summary function
 * "split" by gene (similar to dplyr with attributes)
 * print (similar to data_frame)
 * option to store genotype in sparse matrix format


# snpinfo
 * create a genotype class
    * drop unwanted / unneeded columns
 * improved error checking
 * summary function
 * print (similar to data_frame)
 * data_frame support

 
# gwas_list
 * create gwas_list
    * reduce down to common data
    * warning / stop if nothing in common
 * summary function
 * print (similar to data_frame)
    * number of subjects (first x)
    * number of snps (first x)
    * number of "genes" (first x)
    * number of "chromosomes" (first x)
    * print(pheno)
    * print(geno)
    * print(snpinfo)
 * subsetting (see multitable - Extract.R, mefa, mefa4)
    * [subjects, snps, silevel]
    * `[[ subject/snp/aggregateBy/chr/snpName ==
 * get_phenotype
 * get_snpinfo
 * get_genotype
 * get_kinship
 * get_phenotype_attributes
 * get_snpinfo_attributes
 * get_genotype_attributes
 * get_kinship_attributes
 * option to store genotype in sparse matrix format
 * str.gwas.list <- function(object, give.attr = FALSE, hide.internals = TRUE, ...){

# Apply function for gwas_dataset
 * function(X, MARGIN, FUN, ...) {

# reporting

 * compare p-values of test (facet for > 2 test ???)
 * manhattan plot of p-values
 * summary of inputs / run
 * results to xlsx
 

# misc

# single object functions

 * summary
 * print
 * subsetting
 

# multi object functions

 * intersect phenotype, genotype, snpinfo into a data set to run tests on.
 * gene lookup (return just pheno, geno, snpinfo of that gene)
 * seqMeta parts (score, cov, maf, etc )
 * test <- function(geno, snpinfo, pheno, silevel, fun=, args=)???
 
 
# test

 * gwasrlm with weights
 * burden test (create a wrapper function fo Tx, MB)
 * skat
 * permutation based results
 * Firth (speed up???)